applications/tari_merge_mining_proxy/src/proxy.rs

```diff
@@ -605,20 +642,41 @@ impl InnerService {
             }
         }
 
-        for monerod_url in self.config.monerod_url.iter() {
-            let uri = format!("{}{}", monerod_url, uri.path()).parse::<Url>()?;
-            match reqwest::get(uri.clone()).await {
-                Ok(_) => {
-                    let mut lock = self.last_available_server.write().expect("Write lock should not fail");
-                    *lock = Some(monerod_url.to_string());
-                    info!(target: LOG_TARGET, "Monerod server available: {:?}", uri.clone());
-                    return Ok(uri);
-                },
-                Err(_) => {
-                    warn!(target: LOG_TARGET, "Monerod server unavailable: {:?}", uri);
-                    continue;
-                },
+        let current_url = {
+            let lock = self
+                .last_available_server
+                .read()
+                .expect("Read lock should not fail")
+                .clone();
+            match lock {
+                Some(url) => url,
+                None => "".to_string(),
             }
+        };
+        let mut iter = self.config.monerod_url.iter();
+        let _ = iter.find(|&x| x == &current_url);
```

Comment:
This statement doesn't seem to do anything as the result is swallowed?

Reply:
I amusing this feature of `find`:
``` rust
Stopping at the first true:
let a = [1, 2, 3];

let mut iter = a.iter();

assert_eq!(iter.find(|&&x| x == 2), Some(&2));

// we can still use `iter`, as there are more elements.
assert_eq!(iter.next(), Some(&3));
```

Reply:
ah interesting! maybe just add a comment to that line - something like "advance the iterator to the position of current_url"

Reply:
Ja I would have made this more readable by using the `position` method and then the `skip` method to start from that position in the loop.

Reply:
even better

Reply:
I do not think using `skip` - which is superfluous - will add more value than the comment I have added, as advancing the iterator after the use of `position` and `find` is identical.

```rust
Basic usage:
let a = [1, 2, 3];

let mut iter = a.iter();

assert_eq!(a.iter().position(|&x| x == 2), Some(1));

// we can still use `iter`, as there are more elements.
assert_eq!(iter.next(), Some(&3));

```

Reply:
The point would be to not use a `loop` but rather use the position to find how much to skip and then use a For statement on the iterator that is clearly defined in code using a skip and a chain statement to go through the list a second time.

It's much easier to be explicit on how your are iterating by using iterators and combinators.

Working on your basic example iterating through `a` starting from 2 and then going through the list a second time you would do this:

```Rust
for a.iter().skip(a.iter().position(|x| ==2).chain(a.iter()) {
// do stuff
}
```

Reply:
Re-arranged the list to start from the next unused entry, expanded the iterator with the chained operator, removed the loop.